### ISSUE TITLE
Create S. pombe transcriptome gff and fasta file 

### ISSUE THREAD

**Issue by 3mma-mack** [riboviz/example-datasets#46]
To add a schizosaccharomyces pombe example dataset, the genome gff and fasta files will need to be used to create transcriptome gff and fasta files 

I have found `script_for_transcript_annotation.Rmd` and plan to use this as a starting point. Currently the plan is to create a mini dataset in R using one or two genes from the genome gff and fasta files to test this script and to familiarise myself with the data structures.

**Comment by 3mma-mack** [comment]
To set the framework for creation of a new example dataset and to create a place that will eventuall hold my gff and fasta files I have created branch S_pombe_Duncan-46, and created annotation and contaminants folders with some filler files in them.

**Comment by 3mma-mack** [comment]
I have found Isabel Birds's [code](https://github.com/IsabelBirds/Stuff_to_share/blob/main/Riboviz_work/Converting_fly_genome.Rmd) for creating transcriptome files and have been working to understand the steps, however the script requires bedtools for the extraction of DNA sequences from the fasta files and I am using a windows laptop which does not support bedtools. 

To work around this I am going to combine segment of code from  [script_for_transcript_annotation.Rmd](https://github.com/riboviz/riboviz/blob/master/rmarkdown/script_for_transcript_annotation.Rmd), which does not require bedtools, with the code from Isabel Birds, specifically the lines required for extracting sequences:
```
cds_flank_annot <- unlist(GRangesList(tmp_gff))
# Obtain sequences of CDS+UTRs from genome
cds_flank_seq <- yeast_genome[cds_flank_annot]
names(cds_flank_seq) <- cds_flank_annot$Name
```
After editing the code so that the file names and objects were compatible, the line `cds_flank_seq <- yeast_genome[cds_flank_annot]` produced the error 
```
Error in .subset_by_GenomicRanges(x, i) : 
  'x' must have names when subsetting by a GenomicRanges subscript
```
So I began exploring different potential methods of extracting the DNA sequence and looking at the data as different objects. 
Eventually, one function produced the error:
```
Error in extractTranscriptSeqs(yeast_genome, cds) : 
  the transcripts in 'transcripts' are on chromosomes that are not in 'x'
```
and upon checking the names used for the chromosomes in the processed gff file in Isabel Birds code was different to that in the processed fasta file used in script_for_transcript_annotation.Rmd. So these were changed using 
`names(yeast_genome) <- paste(c("chr_II_telomeric_gap", "I", "II","III", "mating_type_region", "mitochondrial"))`.
Following the name change, `cds_flank_seq <- yeast_genome[cds_flank_annot]` worked, producing a DNAStringSet slightly larger than the original DNAStringSet that was 'yeast_genome'

From here I am going to continue slowly working through the codes and attempt to combine them. It remains to be seen if these will produce the correct files but I am learning a lot along the way.

**Comment by 3mma-mack** [comment]
I am taking a slightly different approach that may or may not be correct, but I have found a fasta file on [Pombase](https://www.pombase.org/data/genome_sequence_and_features/feature_sequences/) that has the sequences for CDS, introns and UTRs. 
I am working on an adaptation of the script_for_transcript_annotations.Rmd code to work with these files and so far I am able to produce another fasta file with [my code](https://github.com/3mma-mack/Stuff_to_share/blob/script_transcript_annot/Riboviz_work/s_pombe_transcriptome_generation.Rmd), which is in a fork of Isabel Bird's branch containing her code for transcriptome file generation, however I still need to do a lot of work to get it running as needed. 

My next steps are:
 
-  Download fasta file containing intron sequences from pombase and remove them from the cds_UTR_intron fasta file if possible, as these may cause some issues with alignment 
- Read over the newly updated [script_for_transcript_annotation](https://github.com/riboviz/riboviz/blob/change-yeast_codon_pos_i200-194/rmarkdown/script_for_transcript_annotation.Rmd), specifically for changes to the section 'Sequences on the negative strand'

**Comment by 3mma-mack** [comment]
I've gone back to Pombase and have found some more fasta files. This time I have found three, one for the CDS sequences, 3'UTRs and 5' UTRs
Both Isabel Birds' Converting_fly_genome and script_for_transcript_annotation seem to set the UTRs at 250 nt, so I looked into the lengths of the UTRs using adapted code provided by @ewallace
 
[S_pombe_3UTR_length.pdf](https://github.com/riboviz/example-datasets/files/6036144/S_pombe_3UTR_length.pdf)
[S_pombe_5UTR_lengths.pdf](https://github.com/riboviz/example-datasets/files/6036146/S_pombe_5UTR_lengths.pdf)

The means for 3'UTRs and 5'UTRs are 427 and 289 nt long respectively, but the maximum lengths are over 4000 nt. So my next step is to decide whether I should edit to UTRs to be a consistent length like those used for saccharomyces or leave them as they are. 

Following that I will need to find any sequences from negative strands using the gff file and create reverse complements. Luckily the sequences are all labelled with their gene IDs, so can be matched with the data in the gff file

**Comment by acope3** [comment]
@3mma-mack Check the new branch `spombe_annot_acope3-46` on example_datasets for S. pombe annotations. I've added the rscript I used to generate these files to the riboviz branch `create_riboviz_style_cds_gff_acope3-278`. As we discussed yesterday, this is a pretty limited script, but could be useful. I'm sure we'll be able to expand the functionality with more of us working on it.

**Comment by 3mma-mack** [comment]
Great, thank you! I'll check those out and see if I can get the Duncan & Mata dataset to run on Eddie

**Comment by 3mma-mack** [comment]
I ran `riboviz.tools.check_fasta_gff` on the the S. pombe annotations in an interactive session on Eddie on the develop branch of riboviz and got the following output:

```
python -m riboviz.tools.check_fasta_gff -f /exports/eddie/scratch/s1734289/riboviz/example-datasets/fungi/schizosaccharomyces/annotation/Schizosaccharomyces_pombe_CDS_w_250utrs.fasta -g /exports/eddie/scratch/s1734289/riboviz/example-datasets/fungi/schizosaccharomyces/annotation/Schizosaccharomyces_pombe_CDS_w_250utrs.gff3

Created by: RiboViz
Date: 2021-02-26 12:16:53.557274
Command-line tool: /exports/eddie/scratch/s1734289/riboviz/riboviz/riboviz/tools/check_fasta_gff.py
File: /exports/eddie/scratch/s1734289/riboviz/riboviz/riboviz/tools/check_fasta_gff.py
Version: commit b75c1ba0e2a541d6f256e2d295b352a3aef969a9 date 2021-02-25 16:16:12+00:00

Configuration:
fasta_file      /exports/eddie/scratch/s1734289/riboviz/example-datasets/fungi/schizosaccharomyces/annotation/Schizosaccharomyces_pombe_CDS_w_250utrs.fasta
gff_file        /exports/eddie/scratch/s1734289/riboviz/example-datasets/fungi/schizosaccharomyces/annotation/Schizosaccharomyces_pombe_CDS_w_250utrs.gff3
start_codons    ['ATG']

Issue summary:
Issue   Count
NoStartCodon    3001
InternalStopCodon       2990
NoStopCodon     2957
IncompleteFeature       1242
DuplicateFeatureId      0
MultipleCDS     0
SequenceNotInFASTA      0
SequenceNotInGFF        0
DuplicateFeatureIds     0
NoIdName        0
```
Having checked my cds.fa.gz DNAStringset object I was working with in RStudio, there are 6998 elements, so 6998 CDS sequences. As riboviz.tools.check_fasta_gff is showing around 3000 with NoStartCodons, InternalStopCodo or NoStopCodon is it possible that these are sequences that were originally on the negative strand @acope3?

[features_issues.txt](https://github.com/riboviz/example-datasets/files/6049770/features_issues.txt)

**Comment by acope3** [comment]
Ugh, I knew I forgot to do something. The code should be handling negative strands. I ran check_fasta_gff on S. cerevisiae CDS and GFF3 file output by this code and did not run into any issues, but I get the same issue as you when running it on S. pombe.

**Comment by acope3** [comment]
Taking a closer look at the the GFF3 file from PomBase, it looks like tRNA genes, rRNA genes, ncRNA genes, etc. also have the CDS label for some reason that is unclear to me. These are obviously not protein-coding sequences, so my guess is these are the cause of our problem.

**Comment by acope3** [comment]
@3mma-mack Okay, so that helped (see attached), but didn't entirely fix the issue. Taking a closer look, I think it is an issue with negative strands, specifically when the gene has an intron. The PomBase GFF3 orders multi-exons genes on the negative strand such that the exon containing the start codon is first. The code we have assumes that if it is on the negative strand, then the exon with the start codon would be last (the code currently uses reverseComplement).
[features_issues.txt](https://github.com/riboviz/example-datasets/files/6050225/features_issues.txt)

**Comment by acope3** [comment]
Yeah, looks like the issue was with handling the multi-exon genes. I'm attaching some updated output from check_fasta_gff after updating the code to handle these pre-ordered exons on the negative strand. As you can see, most of the genes failing are now the non-protein coding sequences. Looks like a few pseudogenes are slipping in, as well. I'll need to update the filtering functionality to handle these cases. 
[features_issues_after_fixing_multi_exon_neg_strand_issue.txt](https://github.com/riboviz/example-datasets/files/6050917/features_issues_after_fixing_multi_exon_neg_strand_issue.txt)

**Comment by 3mma-mack** [comment]
Okay great, that is definitely an improvement! Thanks for keeping me updated and explaining what you're doing

**Comment by 3mma-mack** [comment]
I've been trying to run the [create_files_for_riboviz.R](https://github.com/riboviz/riboviz/blob/create_riboviz_style_cds_gff_acope3-278/rscripts/create_files_for_riboviz.R) code on Rstudio with R version 4.0.3 with the command `system('rscript create_files_for_riboviz.R -i S_pombe.fa.gz -g S_pombe.gff.gz')`and it ran up to "creating riboviz-style CDS..." before producing this error:

```
[1] "Creating riboviz-style CDS..."
Error in .subset_by_GenomicRanges(x, i) : 
  when subsetting by a GenomicRanges subscript, the names of the object
  to subset must contain the seqnames of the subscript
Calls: convertGenomeToCDSFile -> [ -> [ -> .subset_by_GenomicRanges
Execution halted
[1] 1 
```
 
I need to do more investigating and learn how to use the debugger, but if anyone has any ideas that would really help

**Comment by 3mma-mack** [comment]
I have done some more investigating into my earlier problem and noticed that the seqnames of `cds_flank_annot` do not match the names given in the genome file, which was causing the issue with the line` cds_flank_seq <- genome[cds_flank_annot]` 
```
chromosome_labels <- unique(seqnames(cds_flank_annot))
> chromosome_labels
[1] I                    II                   chr_II_telomeric_gap
[4] III                  mitochondrial        mating_type_region  
Levels: chr_II_telomeric_gap I II III mating_type_region mitochondrial
> current.names
[1] "chr_II_telomeric_gap Schizosaccharomyces_pombe"
[2] "I Schizosaccharomyces_pombe"                   
[3] "II Schizosaccharomyces_pombe"                  
[4] "III Schizosaccharomyces_pombe"                 
[5] "mating_type_region Schizosaccharomyces_pombe"  
[6] "mitochondrial Schizosaccharomyces_pombe" 

```
I noticed the function `renameGenome` in the code but I couldn't figure out how to which argument to use to add new names when running the script from the RStudio console so I added the lines:
```
print("Renaming genome...")
genome <- renameGenome(genome, new.names = c('chr_II_telomeric_gap', 'I', 'II', 'III', 'mating_type_region', 'mitochondrial'))
print("done")
```
which calls the function renameGenome and makes the names used in the genome match the seqnames of cds_flank_annot. I added these lines to a copy of the latest version of `create_files_for_riboviz.R`, as I was having the issue when using a copy of the code from both before and after the latest commit. 

The code was successfully run using the command: `system('rscript test_generate_riboviz_files.R')` in RStudio. The inputs and [arguments](https://github.com/riboviz/riboviz/blob/b9489e2e27b0b03abaee48dad1d2cd210c335819/rscripts/create_files_for_riboviz.R#L337) were present in the code so I just kept these as they were. 

However, when I checked the files produced with `check_fasta_gff` there were over 4000 features missing stop codons. 

[features_issues_04_03_2021.txt](https://github.com/riboviz/example-datasets/files/6084307/features_issues_04_03_2021.txt)

Oddly, most of the codons listed in NoStopCodon end in NN, (ie ANN), and the number of NoStopCodons is 1834, exactly the same as the number of NoStopCodons in @acope3's  'features_issues_after_fixing_multi_exon_neg_strand_issue.txt'. I'm not entirely sure what that means, but I can't see where an issue has occured as I tried copying and running the code from `create_files_for_riboviz.R` a few times, including the `renameGenome` lines mentioned above each time, and was getting the same result from `check_fasta_gff` .

**Comment by 3mma-mack** [comment]
It's not a problem, at least I've now learnt to watch out with inconsistent data names and how to fix the issue
 
I ran this with the version from commit b9489e2 which I think is the most recent version of the function. I'll try redownloading the original fasta and gff files, update the chromosome names, and run them again to see if that helps.
It is likely that I've accidentally added a space or something in the wrong place while editing which could be leading to the extra NoStopCodon's, but I'll see how it goes with the next run.

**Comment by acope3** [comment]
Sounds good. For reference, below is the command I used. These updated files have been pushed to the relevant branch for example-datasets.

`Rscript --vanilla rscripts/create_files_for_riboviz.R -i ~/Schizosaccharomyces_pombe_all_chromosomes.fa -g ~/Schizosaccharomyces_pombe_all_chromosomes.gff3 --out_dir /data/cope/example-datasets/fungi/schizosaccharomyces/annotation/ --out_cds Schizosaccharomyces_pombe_CDS_w_250utrs.fasta --out_gff Schizosaccharomyces_pombe_CDS_w_250utrs.gff3 -b 250 --remove_trailing \\: --filter_seq type:mRNA,type:CDS --exons_preordered`

**Comment by 3mma-mack** [comment]
Following a chat with @ewallace  and weekend of unsuccessfully trying to produce files with a similar error level to those produced with Alex, I am going to spot check the files I have been producing to see how the data compares to sequences, and whether problems are conserved to features on a particular strand, to genes with splicing or a combination

As I am running the create_files_for_riboviz.R code without making changes to the code, except changing arguments to reflect my directories, I am unsure of the underlying problem, as it could be anything from a package version to the fact I'm doing this on a windows laptop. Hopefully the spot checks will shed some light on the issue.

**Comment by 3mma-mack** [comment]
I have created subsampled versions of Alex's gff and fasta files, as well as a mini gff from Schizosaccharomyces_pombe_all_chromosomes.gff3 with the same genes which I ran through create_files_for_riboviz.R to generate a mini processed fasta. Each file contained the information for genes originally on the + or - strand, with either one exon or multiple. Using SnapGene, I looked at the individual genes to see what the errors were. 

The genes from the files produced by Alex were all processed as expected, with one CDS that matched the expected sequence from Pombase flanked by 250 nts. The files I produced following a run of create_files_for_riboviz.R had a duplicated CDS, and inconsistently sized UTRs. This is consistent with the [graphs](https://github.com/riboviz/example-datasets/issues/21#issuecomment-792987483) produced by a run of the Duncan and Mata dataset when using the different annotation files.

I have created a document containing observations and images from the SnapGene analysis.

[Analysis of genes in snapgene.docx](https://github.com/riboviz/example-datasets/files/6116854/Analysis.of.genes.in.snapgene.docx)

I am still confused as to why all of my outputs are full of errors but this does mean that the [annotation](https://github.com/riboviz/example-datasets/tree/spombe_annot_acope3-46/fungi/schizosaccharomyces/annotation) files on branch spombe_annot_acope3-46 are as expected.
